### ISSUE TITLE
How could I test my own dataset?

### ISSUE THREAD

**Issue by Chen65010445** [hysts/pytorch_image_classification#16]
I have a folder containing the images which need to evaluate. All images where stored as 
"train dataset folder is:"
/path/female_bag/female_bag*.jpg,
/path/makeup_bag/makeup_bag*.jpg,

" test dataset folder is:"
/path/female_bag/female_bag*.jpg,
/path/makeup_bag/makeup_bag*.jpg,

How should I write the code to test those images?

**Comment by Chen65010445** [comment]
And how can I making a confusion-matrix at the end of the test?

**Comment by hysts** [comment]
Hi, @Chen65010445 

Sorry for the late reply.

I think you can use [this code](https://github.com/hysts/pytorch_image_classification/issues/12#issuecomment-656471334) to run predictions on your test dataset. (You should modify the code to run on batches.)

As for a confusion matrix, you can compute it by using prediction results obtained from the code above and `sklearn.metrics.confusion_matrix`.

Hope it helps.

**Comment by Chen65010445** [comment]
That code was used to test an image how could I run prediction on folders?

**Comment by hysts** [comment]
@Chen65010445 

You can use [`torchvision.datasets.ImageFolder`](https://pytorch.org/docs/stable/torchvision/datasets.html#imagefolder). You can create a test data loader with the following code.

```python
import torch
import torchvision

from torch.utils.data import DataLoader

from pytorch_image_classification import get_default_config, create_transform

config = get_default_config()
config.merge_from_file('/path/to/your/config.yaml')

test_transform = create_transform(config, is_train=False)
test_dataset = torchvision.datasets.ImageFolder('/path/to/your/test/dataset/directory',
                                                transform=test_transform)

test_loader = DataLoader(
    test_dataset,
    batch_size=config.test.batch_size,
    num_workers=config.test.dataloader.num_workers,
    shuffle=False,
    drop_last=False,
    pin_memory=config.test.dataloader.pin_memory)
```

**Comment by Chen65010445** [comment]
Thank you for your help. I will make a try. Is that OK I continue to ask you if I have questions?

**Comment by hysts** [comment]
Sure.

**Comment by Chen65010445** [comment]
Sorry one more question. Where is the config.yaml you mentioned in the code coming from? Is it from the config file I changed before training or the output in checkpoint?

**Comment by hysts** [comment]
@Chen65010445 

You need to prepare a new configuration file from the one you used for training. You need to specify parameters for the test, like `config.test.batch_size`.

**Comment by Chen65010445** [comment]
`import torch
import torchvision
import pytorch_image_classification
import torch.nn.functional as F
import time
import tqdm
import numpy as np

from torch.utils.data import DataLoader
from pytorch_image_classification import get_default_config, create_transform
from pytorch_image_classification.utils import AverageMeter
from sklearn.metrics import confusion_matrix


config = get_default_config()
config.merge_from_file('/home/tmu/pytorch_image_classification/config.yaml')

device = torch.device(config.device)

test_transform = create_transform(config, is_train=False)
test_dataset = torchvision.datasets.ImageFolder('/home/tmu/Downloads/datan/T2/test',
                                                transform=test_transform)

test_loader = DataLoader(
    test_dataset,
    batch_size=config.test.batch_size,
    num_workers=config.test.dataloader.num_workers,
    shuffle=False,
    drop_last=False,
    pin_memory=config.test.dataloader.pin_memory)

model = pytorch_image_classification.create_model(config)
checkpoint = torch.load(config.merge_from_file)

model.load_state_dict(checkpoint['model'])
model.to(device)
model.eval

transform = pytorch_image_classification.create_transform(config, is_train = False)
data = transform(test_dataset)
with torch.no_grad():
    pred = model(data.unsqueeze(0).to(device))
prob = F.softmax(pred, dim = 1).cpu()
    
def evaluate(config, model, test_loader, loss_func, logger):
    device = torch.device(config.device)

    model.eval()

    loss_meter = AverageMeter()
    correct_meter = AverageMeter()
    start = time.time()

    pred_raw_all = []
    pred_prob_all = []
    pred_label_all = []
    with torch.no_grad():
        for data, targets in tqdm.tqdm(test_loader):
            data = data.to(device)
            targets = targets.to(device)

            outputs = model(data)
            loss = loss_func(outputs, targets)

            pred_raw_all.append(outputs.cpu().numpy())
            pred_prob_all.append(F.softmax(outputs, dim=1).cpu().numpy())

            _, preds = torch.max(outputs, dim=1)
            pred_label_all.append(preds.cpu().numpy())

            loss_ = loss.item()
            correct_ = preds.eq(targets).sum().item()
            num = data.size(0)

            loss_meter.update(loss_, num)
            correct_meter.update(correct_, 1)

        accuracy = correct_meter.sum / len(test_loader.dataset)

        elapsed = time.time() - start
        logger.info(f'Elapsed {elapsed:.2f}')
        logger.info(f'Loss {loss_meter.avg:.4f} Accuracy {accuracy:.4f}')

    preds = np.concatenate(pred_raw_all)
    probs = np.concatenate(pred_prob_all)
    labels = np.concatenate(pred_label_all)
    confusion = confusion_matrix(preds,probs)
    return preds, probs, labels, loss_meter.avg, accuracy, confusion
`
Sorry to disturb you againg. I merged part of the evaluate.py and the code you told before. Am I followed your idea? Maybe this is a really really stupid question, but I really need your help.

**Comment by hysts** [comment]
@Chen65010445 

No problem, and yes, you are basically following my idea.
(BTW, I edited your comment a bit because the syntax highlight wasn't working.)

I think some parts are redundant, and there're some typos, but the code should work if you fix them.

**Comment by hysts** [comment]
Um, I think there's a bug in your code to compute a confusion matrix, but the rest is basically ok. I fixed some of the typos in the first part of your code.

```python
config = get_default_config()
config.merge_from_file('path/config.yaml')

device = torch.device(config.device)

test_transform = create_transform(config, is_train=False)
test_dataset = torchvision.datasets.ImageFolder('path/dataset',
                                                transform=test_transform)

test_loader = DataLoader(
    test_dataset,
    batch_size=config.test.batch_size,
    num_workers=config.test.dataloader.num_workers,
    shuffle=False,
    drop_last=False,
    pin_memory=config.test.dataloader.pin_memory)

model = pytorch_image_classification.create_model(config)
checkpoint = torch.load('/path/to/your/model.pth')

model.load_state_dict(checkpoint['model'])
model.to(device)
model.eval()
```

**Comment by hysts** [comment]
And for the computation of the confusion matrix, you need to put ground-truth labels and predicted labels into `confusion_matrix`, instead of the predictions and probabilities.
I modified your code a bit as follows, but I'm not sure if it really works because I current don't have dataset and models to test the function.

```python
def evaluate(config, model, test_loader, loss_func, logger):
    device = torch.device(config.device)

    model.eval()

    loss_meter = AverageMeter()
    correct_meter = AverageMeter()
    start = time.time()

    pred_raw_all = []
    pred_prob_all = []
    pred_label_all = []
    gt_labels = []
    with torch.no_grad():
        for data, targets in tqdm.tqdm(test_loader):
            data = data.to(device)
            targets = targets.to(device)

            outputs = model(data)
            loss = loss_func(outputs, targets)

            pred_raw_all.append(outputs.cpu().numpy())
            pred_prob_all.append(F.softmax(outputs, dim=1).cpu().numpy())

            _, preds = torch.max(outputs, dim=1)
            pred_label_all.append(preds.cpu().numpy())

            gt_labels.append(targets.cpu().numpy())

            loss_ = loss.item()
            correct_ = preds.eq(targets).sum().item()
            num = data.size(0)

            loss_meter.update(loss_, num)
            correct_meter.update(correct_, 1)

        accuracy = correct_meter.sum / len(test_loader.dataset)

        elapsed = time.time() - start
        logger.info(f'Elapsed {elapsed:.2f}')
        logger.info(f'Loss {loss_meter.avg:.4f} Accuracy {accuracy:.4f}')

    preds = np.concatenate(pred_raw_all)
    probs = np.concatenate(pred_prob_all)

    predicted_labels = np.concatenate(pred_label_all)
    gt_labels =np.concatenate(gt_labels)

    confusion = confusion_matrix(gt_labels, predicted_labels)

    return preds, probs, predicted_labels, loss_meter.avg, accuracy, confusion
```

**Comment by Chen65010445** [comment]
Thank you for your help I'll make a try.

**Comment by Chen65010445** [comment]
![选区_041](https://user-images.githubusercontent.com/28993670/92202274-368c2e00-eeb1-11ea-8e1e-3df526f0a5b1.png)
Is this mean I'm loading the wrong local_dict or wrong checkpoint path?

**Comment by hysts** [comment]
@Chen65010445 

Hmm... I have no idea.
I think you should print `model` and `checkpoint['model']` to check if they are what you are trying to use.
